File: python/find_closest_node_to_given_two_nodes.py
from collections import deque, defaultdict


class Solution:
    def closestMeetingNode(self, edges: list[int], node1: int, node2: int) -> int:
        """time complexity: O(n), where n is the number of edges. space complexity: O(n)"""
        graph: dict[int, list[int]] = defaultdict(list)
        for i in range(len(edges)):
            graph[i].append(edges[i])
        node1_dis_map: dict[int, int] = {}
        node2_dis_map: dict[int, int] = {}
        self.bfs(node1, node1_dis_map, graph)
        self.bfs(node2, node2_dis_map, graph)
        result: int = -1
        min_dis: int = float("inf")
        for i in range(len(edges)):
            if i in node1_dis_map and i in node2_dis_map:
                if min_dis > max(node1_dis_map[i], node2_dis_map[i]):
                    result = i
                    min_dis = max(node1_dis_map[i], node2_dis_map[i])
        return result

    def bfs(self, src: int, dis_map: dict[int, int], graph: dict[int, list[int]]):
        queue: deque[tuple[int, int]] = deque([(src, 0)])
        dis_map[src] = 0
        while len(queue) > 0:
            node, dis = queue.popleft()
            for nei in graph[node]:
                if nei not in dis_map:
                    queue.append((nei, dis + 1))
                    dis_map[nei] = dis + 1


def test_closestMeetingNode_case_1():
    # arrange
    edges: list[int] = [2, 2, 3, -1]
    node1: int = 0
    node2: int = 1
    expected: int = 2

    # act
    solution = Solution()
    actual = solution.closestMeetingNode(edges, node1, node2)

    # assert
    assert expected == actual


def test_closestMeetingNode_case_2():
    # arrange
    edges: list[int] = [1, 2, -1]
    node1: int = 0
    node2: int = 2
    expected: int = 2

    # act
    solution = Solution()
    actual = solution.closestMeetingNode(edges, node1, node2)

    # assert
    assert expected == actual

File: rust/src/code/find_closest_node_to_given_two_nodes.rs
use std::cmp::max;
use std::collections::{HashMap, VecDeque};

struct Solution;

impl Solution {
    // NOTE: time complexity O(n) where n is the number of nodes
    pub fn closest_meeting_node(edges: Vec<i32>, node1: i32, node2: i32) -> i32 {
        let mut graph: HashMap<i32, Vec<i32>> = HashMap::new();
        for i in 0..edges.len() {
            if edges[i] == -1 {
                graph.entry(i as i32).or_insert(vec![]);
            } else {
                graph.entry(i as i32).or_insert(vec![edges[i]]);
            }
        }
        let mut node1_map: HashMap<i32, i32> = HashMap::new();
        let mut node2_map: HashMap<i32, i32> = HashMap::new();
        Self::bfs(node1, &mut node1_map, &graph);
        Self::bfs(node2, &mut node2_map, &graph);

        let mut result: i32 = -1;
        let mut min_dis: i32 = i32::MAX;
        for i in 0..edges.len() {
            let node: i32 = i as i32;
            if node1_map.contains_key(&node) && node2_map.contains_key(&node) {
                if min_dis > max(node1_map[&node], node2_map[&node]) {
                    min_dis = max(node1_map[&node], node2_map[&node]);
                    result = i as i32;
                }
            }
        }
        result
    }

    fn bfs(node: i32, node_map: &mut HashMap<i32, i32>, graph: &HashMap<i32, Vec<i32>>) {
        let mut queue: VecDeque<(i32, i32)> = VecDeque::from([(node, 0)]);
        node_map.insert(node, 0);
        while let Some((n, dis)) = queue.pop_front() {
            for neighbor in graph[&n].iter() {
                if node_map.contains_key(neighbor) {
                    continue;
                }
                node_map.insert(*neighbor, dis + 1);
                queue.push_back((*neighbor, dis + 1));
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_closest_meeting_node_case_1() {
        // arrange
        let edges: Vec<i32> = vec![2, 2, 3, -1];
        let node1: i32 = 0;
        let node2: i32 = 1;
        let expected: i32 = 2;

        // act
        let actual = Solution::closest_meeting_node(edges, node1, node2);

        // arrange
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_closest_meeting_node_case_2() {
        // arrange
        let edges: Vec<i32> = vec![1, 2, -1];
        let node1: i32 = 0;
        let node2: i32 = 2;
        let expected: i32 = 2;

        // act
        let actual = Solution::closest_meeting_node(edges, node1, node2);

        // arrange
        assert_eq!(expected, actual);
    }
}
